### PR TITLE
Context rules with fine-grained control of how deep to copy

### PR DESCRIPTION
**Pull request by oflatt**
This PR improves the context rules, with fine-grained control of how deep to copy.
It also fixes our saturation issues by keeping track of a list of eclasses already visited, and subsuming rows that have encountered a cycle.

### CODE REVIEW

File: tree_in_context/src/utility/in_context.egg [hunk 1]
```diff
@@ -1,146 +1,219 @@
...
+       (= scope (Full)))
+      ((let new-inputs
+         (DoAddContext seen ctx scope inputs))
+       (let newpath
+         (PathCons (DoWhile inputs outputs) lhs seen))
+       (union lhs
+         (DoWhile new-inputs
+           (DoAddContext newpath (InLoop new-inputs outputs) scope outputs))))
+       :ruleset context)
+
```
Comment (ezrosent) [inline tree_in_context/src/utility/in_context.egg:216]: Probably don't need all these extra lines

File: tree_in_context/src/utility/in_context.egg [hunk 2]
```diff
@@ -1,146 +1,219 @@
...
+         (Single (DoAddContext seen ctx scope c1))
+         :ruleset context-helpers)
+(rewrite (DoAddContext seen ctx scope (Concat order c1 c2))
+         (Concat order
+           (DoAddContext seen ctx scope c1)
+           (DoAddContext seen ctx scope c2))
+         :ruleset context-helpers)
+
+;; Replace existing contexts.
+;; This relies on the fact that new contexts must be more specific than old ones.
```
Comment (ezrosent) [inline tree_in_context/src/utility/in_context.egg:158]: I know we discussed this in a meeting at some point, but have we written down why this is true anywhere (and continues to be true, say, when we add some of these contexts ahead of time in the Rust pass?)

**Review by ezrosent — APPROVED** [review]

File: tree_in_context/src/utility/in_context.egg (hunk 2 above)
Comment (oflatt) [reply tree_in_context/src/utility/in_context.egg:158]: That's a great question, and I hope to answer that soon by updating the semantics in the repository.
A high-level answer is that we maintain this invariant:
A context node defines the set of values that a program is valid on, and this program is *guaranteed to execute* on these values. That means nested context nodes must pass for all values the program is executed on.

File: tree_in_context/src/utility/in_context.egg [hunk 3]
```diff
@@ -1,146 +1,219 @@
...
+;; A context path is a list of (eclass, eclass) pairs.
+;; In substitution, the path is used to store every context-related eclass we have visited, 
+;; and the corresponding expression at that point.
+;; When we visit an eclass for the second time, we have found a cycle in the egraph.
+;; Then the corresponding eclasses are made equivalent.
+(datatype ContextPath
+  (PathNil)
+  (PathCons Expr Expr ContextPath))
+
+(relation Path-contains (ContextPath ContextPath))
```
[the remark this answers is not in the document]
Comment (oflatt) [reply tree_in_context/src/utility/in_context.egg:30]: added a description with example

**Pull request merged**